File: LeetCode/Explore/TopInterview/Blind75/Tree/230_kth_smallest_element_in_a_bst.cpp
// Recursive Inorder Traversal

// Complexity Analysis
// Time complexity : O(N)O(N) to build a traversal.
// Space complexity : O(N)O(N) to keep an inorder traversal.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  
private:
  vector<int> vp;
  void inOrder(TreeNode *root) {
     if (!root) {
      return;
    }
    inOrder(root->left);
    vp.push_back(root->val);
    inOrder(root->right);
  }
  
public:
  int kthSmallest(TreeNode* root, int k) {
    inOrder(root);
    return vp[k - 1];
  }
};

File: LeetCode/Explore/TopInterview/Blind75/Tree/230_kth_smallest_element_in_a_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  
    def inOrder(self, root: Optional[TreeNode], ans: List[int]):
      if root == None:
        return ans
      self.inOrder(root.left, ans)
      ans.append(root.val)
      self.inOrder(root.right, ans)
  
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
      ans = list()
      self.inOrder(root, ans)
      return ans[k - 1]
